### PR TITLE
Override the YAML constructor for timestamps to output as strings and not datetime objects

### PR DESCRIPTION
**Pull request by brysontyrrell**
Addresses an issue with `--use-json` throws "TypeError: Object of type date is not JSON serializable" if `AWSTemplateFormatVersion` value is not quoted.

#### Which issue(s) does this change fix?

#2476 #4735 

#### Why is this change necessary?

If the value for `AWSTemplateFormatVersion: 2010-09-09` is not in quotes the YAML parser will interpret it as a date and output a datetime object which will throw a `TypeError` when the `--use-json` argument is used.

#### How does it address the issue?

This adds an additional constructor to `yaml_parse()` that forces any `tag:yaml.org,2002:timestamp` to output to a string.

#### What side effects does this change have?

Date processing doesn't appear to be in use internally in SAM so there does not appear to be any other impact.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [ ] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by sriram-mv** [comment]
Thanks for your contribution! would you mind adding unit tests as well?

**Comment by brysontyrrell** [comment]
> Thanks for your contribution! would you mind adding unit tests as well?

Unit test added.

File: samcli/yamlhelper.py [hunk 1]
```diff
@@ -118,6 +118,9 @@ def yaml_parse(yamlstr) -> Dict:
         # json parser.
         return cast(Dict, json.loads(yamlstr, object_pairs_hook=OrderedDict))
     except ValueError:
+        yaml.constructor.SafeConstructor.yaml_constructors[
+            "tag:yaml.org,2002:timestamp"
```
Comment (jfuss) [inline samcli/yamlhelper.py:122]: Nit: Can we make this a Constant like we did for `TAG_STR`?

File: tests/unit/test_yamlhelper.py [hunk 2]
```diff
@@ -161,3 +163,22 @@ def test_unroll_yaml_anchors(self):
...
+            "      Handler: main.handler\n"
+            "      Runtime: python3.7\n"
+            "      CodeUri: .\n"
+            "      Timeout: 600\n"
+        )
+
+        output = yaml_parse(input_template)
+        # Raises a `TypeError` if an unquoted `AWSTemplateFormatVersion` value has been parsed to a
+        # `datetime` object and not a string by `yaml_parse` when using `--use-json` argument.
+        json.dumps(output)
```
Comment (jfuss) [inline tests/unit/test_yamlhelper.py:184]: Nit: Can we validate the expected date is converted to a string? Not throwing an error is probably enough but would like to ensure we are at least asserting the behavior.

**Review by jfuss — CHANGES_REQUESTED** [review]
Mostly just Nits/small asks. 

The one question that came up for me is how does CloudFormation handle this? For example, if I had a random date somewhere in my template, does CloudFormation convert it to string always or lets yaml cover to date or something more conditionally?

I ask because when I was thinking about the test, I realized this may convert all dates to a string and wonder if that is the right thing. Just want to make sure we don't convert something incorrectly and have a wider impact of this change. Generally I think it's good, just want to ensure we understand completely to avoid having to revert later on due to something unexpected. :)

**Comment by brysontyrrell** [comment]
> The one question that came up for me is how does CloudFormation handle this? For example, if I had a random date somewhere in my template, does CloudFormation convert it to string always or lets yaml cover to date or something more conditionally?
> 
> I ask because when I was thinking about the test, I realized this may convert all dates to a string and wonder if that is the right thing. Just want to make sure we don't convert something incorrectly and have a wider impact of this change. Generally I think it's good, just want to ensure we understand completely to avoid having to revert later on due to something unexpected. :)

You're correct that this change casts all dates to strings. I don't know how CloudFormation handles this, but dates as an object or type don't exist in it as far as I know.

**Comment by jfuss** [comment]
> You're correct that this change casts all dates to strings. I don't know how CloudFormation handles this, but dates as an object or type don't exist in it as far as I know.

That is probably the right thing here, just being generally cautious. I can't find any docs or details on this sadly. Which means we either need to test the behavior or ask someone on the team (which would be on us).

**Review by jfuss — APPROVED** [review]

**Review by sriram-mv — APPROVED** [review]
Thanks for adding the tests! Appreciate it :D

**Pull request merged**